index.html

```diff
@@ -1,453 +1,585 @@
+          <code>type</code>
+        </li>
+        <li>Creates an empty array <code>types</code> of <code>TypedDataField</code> to collect all the fields</li>
+        <li>Canonicalizes the input document using the <a href="#dfn-canonicalization-algorithm">canonicalization
+            algorithm</a></li>
+        <li>If `primaryType` is not provided, set `primaryType = "Document"` else use the provided value.</li>
+        <li>For each property in the canonicalized input document, the algorithm checks the type of the value specific
+          to the
+          implementation language.</li>
+        <li>If the type of the value is a primitive <code>boolean</code>, <code>number</code> or <code>string</code>,
+          push an object to <code>types</code> with the <code>name</code> set to the property name of the input
+          document, and <code>type</code> set to the corresponding EIP712 primitive type
+          <ol class="algorithm">
+            <li><code>boolean</code> - maps to <code>bool</code></li>
+            <li><code>number</code> - maps to <code>uint256</code></li>
+            <li><code>string</code> - maps to <code>string</code></li>
+          </ol>
+        </li>
+        <li>If the type of the value is an array, ensure each element of the array has the same primitive type. Push an
+          object to <code>types</code> with the <code>name</code> set to the property name of the input document and
+          <code>type</code> set to the corresponding EIP712 array type
+          <ol class="algorithm">
+            <li><code>boolean[]</code> - maps to <code>bool[]</code></li>
+            <li><code>number[]</code> - maps to <code>uint256[]</code></li>
+            <li><code>string[]</code> - maps to <code>string[]</code></li>
+          </ol>
+          <p class="example" title="Arrays of structs are not yet supported">
+            WARNING: The current algorithm definition does not support auto generating types for arrays of structs. We
+            need
+            to
+            work on that.
+          </p>
+        </li>
+        <li>If the type of the value is an object, call the function recursively on the inner object, and set the return
+          value equal to <code>_recursiveOutput</code>.
+          <ol class="algorithm">
+
+            <li>Set <code>_recursiveTypes = _recursiveOutput[primaryType]</code></li>
+            <li>Push an object to <code>types</code> with the <code>name</code> set to the property name of the input
+              document and <code>type</code> set to the CapitalCased property name - <code>propertyType</code>
+            </li>
+            <li>
+              Set <code>output[propertyType] = _recursiveTypes</code>
+            </li>
+            <li>Loop over <code>_recursiveOutput</code>, and if any keys other than <code>primaryType</code> are
+              present,
+              add them directly to <code>output</code></li>
```

Comment:
This algorithm would produce the same type name if objects appear under the same property name in different objects.
For example, the following two objects I think would both result in duplicate definitions for a "Foo" type (and the second one would be recursive too):

```json
{
  "a": {
    "foo": {}
  },
  "b": {
    "foo": {}
  }
}
```
```json
{
  "foo": {
    "foo": {}
  }
}
```
This can be handled as an error condition:
```suggestion
              add them directly to <code>output</code>. If any such key already has an entry in <code>output</code>, raise an error.</li>
```

Reply:
This is part of the issue of handling duplicate Struct types, and also relates to handling arrays of structs. I am working on a fix for this instead of defaulting to an error.